Fix #989: Crash on Profile edit page landscape.

anandwana001 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
In my earlier PR https://github.com/oppia/oppia-android/pull/1446, I had solved this issue using `retainInstance` which after discussing with the team I get to know it create memory leaks issue and it's better to solve through Injecting ViewModel directly to the presenter.  

<img width="1440" alt="Screenshot 2020-09-02 at 17 24 42" src="https://user-images.githubusercontent.com/25057618/91980835-851cb980-ed45-11ea-9732-fd2c950fdecf.png">
<img width="1440" alt="Screenshot 2020-09-02 at 17 22 19" src="https://user-images.githubusercontent.com/25057618/91980846-8b129a80-ed45-11ea-83af-64613865b1f5.png">


## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

anandwana001 committed:
solving through viewmodel

BenHenning commented on app/src/main/java/org/oppia/app/profileprogress/ProfileProgressViewModel.kt:
```diff
@@ -34,7 +34,7 @@ class ProfileProgressViewModel @Inject constructor(
   private val topicListController: TopicListController,
   private val logger: ConsoleLogger,
   @StoryHtmlParserEntityType private val entityType: String
-) : ObservableViewModel() {
```
Does it just not need to be observable?

rt4914 replied:
@BenHenning Though it is assigned to Akshay, I am actually not able to understand what do you mean by this comment?

BenHenning replied:
Ah. I was confused about why this was removed, but it makes sense now.

BenHenning reviewed (changes requested):
Thanks @anandwana001. Can you confirm there's already a test to verify that the crash isn't occurring?

rt4914 reviewed (changes requested):
Just add a test case which follows the same steps mentioned in the issue:
`Change profile image in portrait mode, now change orientation and click on edit image.`

anandwana001 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into profile-progress-landscape-viewmodel

anandwana001 committed:
added test case and nit change

rt4914 reviewed (approved):
LGTM

BenHenning commented on app/src/sharedTest/java/org/oppia/app/profileprogress/ProfileProgressFragmentTest.kt:
```diff
@@ -218,6 +218,42 @@ class ProfileProgressFragmentTest {
+          R.id.profile_edit_image
+        )
+      ).perform(click())
+      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
```
Please add a comment to the effect: “dialog should still be open after configuration change”

BenHenning replied:
Committed directly.

BenHenning reviewed (approved):
Thanks! Lgtm—feel free to resolve my comment after addressing.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/profileprogress/ProfileProgressFragmentTest.kt:
```diff
@@ -218,6 +218,42 @@ class ProfileProgressFragmentTest {
+          R.id.profile_edit_image
+        )
+      ).perform(click())
+      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
```
```suggestion
      // The dialog should still be open after a configuration change.
      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
```

BenHenning committed:
Update app/src/sharedTest/java/org/oppia/app/profileprogress/ProfileProgressFragmentTest.kt

Add suggested comment.

BenHenning committed:
Merge branch 'develop' into profile-progress-landscape-viewmodel

BenHenning commented:
@rt4914 can you verify the latest changes I made to the branch?

BenHenning commented on app/src/sharedTest/java/org/oppia/app/profileprogress/ProfileProgressFragmentTest.kt:
```diff
@@ -248,6 +248,43 @@ class ProfileProgressFragmentTest {
+    )
+    val activityResult = createGalleryPickActivityResultStub()
+    intending(expectedIntent).respondWith(activityResult)
+    launch<ProfileProgressActivity>(createProfileProgressActivityIntent(internalProfileId)).use {
+      waitForTheView(withText("Admin"))
+      onView(
+        atPositionOnView(
+          R.id.profile_progress_list,
+          0,
+          R.id.profile_edit_image
+        )
+      ).perform(click())
+      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
+        .check(matches(isDisplayed()))
+      onView(withText(R.string.profile_picture_edit_alert_dialog_choose_from_library))
+        .perform(click())
+      intended(expectedIntent)
+      onView(isRoot()).perform(orientationLandscape())
+      intended(expectedIntent)
+      onView(
+        atPositionOnView(
+          R.id.profile_progress_list,
+          0,
+          R.id.profile_edit_image
+        )
+      ).perform(click())
+      // The dialog should still be open after a configuration change.
+      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
+        .check(matches(isDisplayed()))
+    }
```
```suggestion
    launch<ProfileProgressActivity>(createProfileProgressActivityIntent(internalProfileId)).use {
      testCoroutineDispatchers.runCurrent()
      waitForTheView(withText("Admin"))
      onView(
        atPositionOnView(
          R.id.profile_progress_list,
          0,
          R.id.profile_edit_image
        )
      ).perform(click())
      testCoroutineDispatchers.runCurrent()
      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
        .check(matches(isDisplayed()))
      onView(withText(R.string.profile_picture_edit_alert_dialog_choose_from_library))
        .perform(click())
      testCoroutineDispatchers.runCurrent()
      intended(expectedIntent)
      onView(isRoot()).perform(orientationLandscape())
      testCoroutineDispatchers.runCurrent()
      intended(expectedIntent)
      onView(
        atPositionOnView(
          R.id.profile_progress_list,
          0,
          R.id.profile_edit_image
        )
      ).perform(click())
      testCoroutineDispatchers.runCurrent()
      // The dialog should still be open after a configuration change.
      onView(withText(R.string.profile_progress_edit_dialog_title)).inRoot(isDialog())
        .check(matches(isDisplayed()))
    }
```

BenHenning replied:
Added runCurrent calls since merging with develop broke the test.

BenHenning committed:
Update app/src/sharedTest/java/org/oppia/app/profileprogress/ProfileProgressFragmentTest.kt

Add runCurrent synchronization points since after merging with develop this test now uses paused dispatchers by default.

BenHenning commented:
@rt4914 please take a look at the newest commit as well since I needed to fix the new test being added.

rt4914 reviewed (approved):
LGTM, thanks.

BenHenning commented:
Thanks for checking!

BenHenning commented:
Requesting a CP on this since it's a straightforward fix and it's nice to include it in alpha.
